File: Python_codes/LINKEDLIST/singly_linked_list.py
"""
@date : 12/05/2024
@session :  10
@motive : to implement and test singly linked list datastructure
"""

class  Node:
        """
        
        """
        def __init__(self,new_data:any):
            self.data = new_data
            self.next = None
        
        


class singlyLinkedList:
    """
    
    """
    def __init__(self):
          self.head_node = Node(None)
    

    def insert_start(self,new_data)->bool:
       new_node =  Node(new_data) # create new node
       new_node.next =  self.head_node.next # assign to new node next head node next 
       self.head_node.next =  new_node # assign to existing head node next new node

    def insert_end(self,new_data)-> bool:
        new_node = Node(new_data)
        run = self.head_node
        while run.next is not None :
                run = run.next
        new_node.next = run.next # although it None here in other list here might be different
        run.next = new_node
    
    def insert_after(self,existing_data,new_data)->bool:
        run  =  self.head_node.next
        while run != None :
               if run.data == existing_data :
                      break
               run  =  run.next
        else:
               raise ValueError(f"{existing_data} not found")

        new_node =  Node(new_data)
        new_node.next = run.next
        run.next = new_node
       
    def insert_before(self,existing_data,new_data)->bool:
        run = self.head_node
        while run.next != None :
               if run.next.data == existing_data:
                      break
               run = run.next
        else:
               raise ValueError(f"{existing_data} not found")
        # create new node
        new_node = Node(new_data)
        new_node.next = run.next
        run.next = new_node
        return True
    
    
    def get_start(self):
        if self.head_node.next is None :
                raise ValueError("Can not get the start as list is empty")
        return self.head_node.next.data
  
    def get_end(self):
        if self.head_node.next is None :
                raise ValueError("Can not get the start as list is empty")
        run =  self.head_node
        while  run.next is not None:
                run = run.next
        return run.data

    def pop_start(self):
        if self.head_node.next is None : 
                raise ValueError("list is empty!!")
        data = self.head_node.next.data 
        self.head_node.next =self.head_node.next.next
        return data
    
    def pop_end(self):
        if self.head_node.next is None:
              raise ValueError("list is empty !!!")
        run =  self.head_node
        while  run.next.next is not None:
              run = run.next
        data = run.next.data
        run.next = run.next.next
        return data

    def remove_start(self):
        if self.head_node.next is None : 
                raise ValueError("list is empty!!")
        self.head_node.next =self.head_node.next.next

    
    def remove_end(self):
        if self.head_node.next is None:
         raise ValueError("list is empty !!!")
        run =  self.head_node
        while  run.next.next is not None:
              run = run.next
        run.next = run.next.next

    def remove_data(self,new_data):
        if self.head_node.next is None:
         raise ValueError("list is empty !!!")
        run =  self.head_node
        while  run.next is not None and run.next.data != new_data:
              run = run.next
        run.next = run.next.next


    def show(self,s:str):
        if type(s) != str:
              raise TypeError("arg not a string")
        print(s)    
        print("[START] ->",end= ' ')
        trav =  self.head_node.next
        while trav is not None:
                print(f"{trav.data} ->",end=' ')
                trav=trav.next
        print("[END]")

   


def main():
        L = singlyLinkedList()
        print("L.__dict__ : ",L.__dict__)
        print("L.head_node.__dict__ : ",L.head_node.__dict__)
        L.insert_start(10)
        L.insert_start(20)
        L.insert_after(20,15)
        L.insert_end(60)
        L.insert_end(70)
        L.insert_end(80)
        L.insert_end(101)
        L.insert_before(15,45)
        L.show("after adding data")
        data1 = L.get_start()
        data2 = L.get_end()
        L.show("after get data")
        print(f"get data ata start {data1}")
        print(f"get data ata start {data2}")
        data3 = L.pop_start()
        L.show("after pop_start() data")
        data4 = L.pop_end()
        L.show("after pop_end() data")
        L.remove_end()
        L.show("after L.remove_end() data")
        L.remove_data(60)
        L.show("after L.remove_data(15) data")
        L.remove_start()
        L.show("after L.remove_start() data")

main()

File: Java_codes/Linked List/SinglyLinkedList.java
package com.tms.dsa.linkedlist;

class Node {
	public Integer data;
	public Node next;

	public Node(int data) {
		this.data = data;
		this.next = null;
	}
//
//	public int getData() {
//		return data;
//	}
//
//	public void setData(int data) {
//		this.data = data;
//	}
//
//	public Node getNext() {
//		return next;
//	}
//
//	public void setNext(Node node) {
//		this.next = node;
//	}
//	
//
//	public String toString() {
//		return data + " <-> ";
//	}
}

public class SinglyLinkedList {
	Node headNode;

	public SinglyLinkedList() {
		this.headNode = new Node(-1);
	}

	public boolean insertAtStart(int data) {
		Node newNode = new Node(data);
		newNode.next = this.headNode.next;
		this.headNode.next = newNode;
		return true;
	}

	public void insertAtEnd(int data) {
		Node newNode = new Node(data);
		Node traversingNode = this.headNode;
		while (traversingNode.next != null) {
			traversingNode = traversingNode.next;
		}
		newNode.next = traversingNode.next;
		traversingNode.next = newNode;

	}

	public void insertAfter(int existingData, int data) {
		Node run = this.headNode.next;
		while (run != null) {
			if (run.data == existingData)
				break;
			run = run.next;
		}
		if (run == null)
			throw new DataNotFoundException("existing data not found!");
		Node newNode = new Node(data);
		newNode.next = run.next;
		run.next = newNode;

	}

	public void insertBefore(int existingData, int data) {
		Node run = this.headNode;
		while (run.next != null) {
			if (run.next.data == existingData) {
				break;
			}
			run = run.next;
		}
		if (run == null)
			throw new DataNotFoundException("existing data not found!");
		Node newNode = new Node(data);
		newNode.next = run.next;
		run.next = newNode;

	}

	public String toString() {
		String s = "";
		Node run = this.headNode.next;
		s = "START ->";
		while (run != null) {
			s += "(" + run.data + ")->";
			run = run.next;
		}
		s += "END";

		return s;
	}

	public int getStart() {
		if (this.headNode.next == null) {
			throw new EmptyListException("Can not get start of empty list");
		}
		return this.headNode.next.data;
	}

	public int getEnd() {
		if (this.headNode.next == null) {
			throw new EmptyListException("Can not get end of empty list");
		}
		Node run = this.headNode;
		while (run.next != null) {
			run = run.next;
		}
		return run.data;

	}

	public int popStart() {
		if (this.headNode.next == null) {
			throw new EmptyListException("Can not pop start of empty list");
		}
		int data = this.headNode.next.data;
		this.headNode.next = this.headNode.next.next;
		return data;
	}

	public int popEnd() {
		if (this.headNode.next == null) {
			throw new EmptyListException("Can not pop end of empty list");
		}
		Node run = this.headNode;
		while (run.next.next != null) {
			run = run.next;
		}
		int data = run.next.data;
		run.next = run.next.next;
		return data;
	}

	public void removeStart() {
		if (this.headNode.next == null) {
			throw new EmptyListException("Can not remove from empty list");
		}
		this.headNode.next = this.headNode.next.next;
	}

	public void removeEnd() {
		if (this.headNode.next == null) {
			throw new EmptyListException("Can not remove from end for empty list");
		}
		Node r = this.headNode;
		while (r.next.next != null) {
			r = r.next;
		}
		r.next = r.next.next;
	}

	public void removeData(int dataToRemove) {
		if (this.headNode.next == null) {
			throw new EmptyListException("Can not remove data from empty list");
		}
		Node run = this.headNode;
		while (run.next != null && run.next.data != dataToRemove)
			run = run.next;

		if (run.next != null) {
			run.next = run.next.next;
		}
	}

	public boolean find(int data) {
		Node run = this.headNode.next;
		while (run != null && run.data != data) {
			run = run.next;
		}
		return run != null ? true : false;
	}

	public int length() {
		int cnt = 0;
		Node run = this.headNode.next;
		while (run != null) {
			cnt++;
			run = run.next;
		}
		return cnt;
	}

	public SinglyLinkedList concat(SinglyLinkedList other) {
		/*
		 * 1) add this list to new list 2) then add other list to new list using insert
		 * at end for both 1 n 2
		 * 
		 */
		SinglyLinkedList newList = new SinglyLinkedList();
		Node run = this.headNode.next;
		while (run != null) {
			newList.insertAtEnd(run.data);
			run = run.next;
		}
		run = other.headNode.next;
		while (run != null) {
			newList.insertAtEnd(run.data);
			run = run.next;
		}
		return newList;
	}

	public void append(SinglyLinkedList other) {

		if (other.headNode.next == null)
			throw new EmptyListException("Other list is empty");

		Node run = this.headNode;
		while (run.next != null) {
			run = run.next;
		}
		run.next = other.headNode.next;
		other.headNode = null;

	}

	public SinglyLinkedList getReversedList() {
		if (this.headNode.next == null)
			throw new EmptyListException("Other list is empty");
		SinglyLinkedList list = new SinglyLinkedList();
		Node run = this.headNode.next;
		while (run != null) {
			list.insertAtStart(run.data);
			run= run.next;
		}
		return list;

	}

	public void reverseList() {
		if(this.headNode.next == null && this.headNode.next.next ==null) {
			throw new EmptyListException("EMPTY OR ONE ELEMENT LIST CAN NOT BE REVERSED");		
		}
		Node originalFirst =  this.headNode.next;
		Node run =  this.headNode.next.next;
		while(run != null) {
			Node r_next = run.next;
			run.next = this.headNode.next;
			this.headNode.next=run;
			run = r_next;
		}
		originalFirst.next = null;
		
	}

}
